vllm/v1/spec_decode/eagle.py

```diff
@@ -387,14 +388,27 @@ def propose(
                 positions += 1
                 exceeds_max_model_len = positions >= self.max_model_len
                 clamped_positions = torch.where(exceeds_max_model_len, 0, positions)
+            # when enable use_async_scheduling, we shouldn't use in place
+            # operations in case they are modified in next step `prepare_input`
+            # of main model.
+            if self.use_async_scheduling:
+                # Increment the sequence lengths.
+                common_attn_metadata.seq_lens = common_attn_metadata.seq_lens + 1
+                common_attn_metadata.seq_lens_cpu = (
+                    common_attn_metadata.seq_lens_cpu + 1
+                )
+                # For the requests that exceed the max model length, we set the
+                # sequence length to 1 to minimize their overheads in attention.
 
-            # Increment the sequence lengths.
-            common_attn_metadata.seq_lens += 1
-            common_attn_metadata.seq_lens_cpu += 1
-            # For the requests that exceed the max model length, we set the
-            # sequence length to 1 to minimize their overheads in attention.
+                common_attn_metadata.seq_lens.masked_fill(exceeds_max_model_len, 1)
+            else:
+                # Increment the sequence lengths.
+                common_attn_metadata.seq_lens += 1
+                common_attn_metadata.seq_lens_cpu += 1
+                # For the requests that exceed the max model length, we set the
+                # sequence length to 1 to minimize their overheads in attention.
 
-            common_attn_metadata.seq_lens.masked_fill_(exceeds_max_model_len, 1)
+                common_attn_metadata.seq_lens.masked_fill_(exceeds_max_model_len, 1)
```

Comment:
So the race condition is related to the `seq_lens_cpu` tensor?

I don't think anything should be needed apart from to clone this at the right place if so (or ensure a copy is made via other means e.g. out-of-place op.)

In particular I don't think any change to the GPU tensors should be needed if they are only accessed in the main cuda stream (e.g. common_attn_metadata.seq_lens).